### PR TITLE
Implemented a first version to allow for custom removal scripts & retry

### PR DESCRIPTION
**Pull request by MrMCake**
# Change

- Implemented a first version to allow for custom removal scripts & retry
- Moved the removal scripts to utilities to enable reference by non-GitHub pipelines & custom execution
- Cleaned up pipelines & added reference to custom removal for vWan & netApp
- Refactored removal scripts to align approaches & reduce code

Tested multiple times with
- [Compute: Virtualmachines](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.virtualmachines.yml)
- [NetApp: Netappaccounts](https://github.com/Azure/ResourceModules/actions/workflows/ms.netapp.netappaccounts.yml)
- [Resources: Resourcegroups](https://github.com/Azure/ResourceModules/actions/workflows/ms.resources.resourcegroups.yml)
- [Network: Virtualwans](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.virtualwans.yml)

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: utilities/resourceRemoval/Remove-vWan.ps1 [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        [string] $moduleName = 'virtualWans',
+
+        [Parameter(Mandatory = $false)]
+        [string] $ResourceGroupName = 'validation-rg'
+    )
+
+    #Identifying resources
+    $retryCount = 1
+    $retryLimit = 40
+    $retryWaitInSeconds = 15
```
Comment (MariusStorhaug) [inline utilities/resourceRemoval/Remove-vWan.ps1:34]: Suggestion: Put these as parameters with default values.
Wont do it in the suggestion, but need to also fix casing on params and documentation on functions.

```suggestion
    [Cmdletbinding(SupportsShouldProcess)]
    param(
        [Parameter(Mandatory = $false)]
        [string] $moduleName = 'virtualWans',

        [Parameter(Mandatory = $false)]
        [string] $ResourceGroupName = 'validation-rg',
        
        [Parameter(Mandatory = $false)]
        [int] $retryLimit = 40,
        
        [Parameter(Mandatory = $false)]
        [int] $retryWaitInSeconds = 15
    )

    #Identifying resources
    $retryCount = 1
```

File: utilities/resourceRemoval/Remove-NetAppAccount.ps1 [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        [Parameter(Mandatory = $false)]
+        [string] $moduleName = 'netAppAccounts',
+
+        [Parameter(Mandatory = $false)]
+        [string] $ResourceGroupName = 'validation-rg'
+    )
+
+    $retryCount = 1
+    $retryLimit = 40
+    $retryWaitInSeconds = 15
```
Comment (MariusStorhaug) [inline utilities/resourceRemoval/Remove-NetAppAccount.ps1:14]: Suggestion: Put these as parameters with default values.
Wont do it in the suggestion, but need to also fix casing on params and documentation on functions.

Not for now: I think the retry mechanism should be outside of this script. I'm not sure it should not have to be up to each script to handle retries. Thinking more the script should have logic to handle a removal, if it throws an error it returns to caller, and caller handles the retry by calling the script again. But there are probably reasons to have retries in these scripts too I can imagine.

```suggestion
    [Cmdletbinding(SupportsShouldProcess)]
    param(
        [Parameter(Mandatory = $false)]
        [string] $moduleName = 'netAppAccounts',

        [Parameter(Mandatory = $false)]
        [string] $ResourceGroupName = 'validation-rg',
        
        [Parameter(Mandatory = $false)]
        [int] $retryLimit = 40,
        
        [Parameter(Mandatory = $false)]
        [int] $retryWaitInSeconds = 15
    )

    $retryCount = 1
```

File: utilities/resourceRemoval/Remove-vWan.ps1 [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    #Building array of resources to remove (in the required order)
+    $resourcesToRemove = @()
+    if ($vpnGatewayResourceId) { $resourcesToRemove += Get-AzResource -ResourceId $vpnGatewayResourceId }
+    if ($virtualHubResourceId) { $resourcesToRemove += Get-AzResource -ResourceId $virtualHubResourceId }
+    if ($vpnSiteResourceId) { $resourcesToRemove += Get-AzResource -ResourceId $vpnSiteResourceId }
+    if ($vWANResourceId) { $resourcesToRemove += Get-AzResource -ResourceId $vWANResourceId }
+
+    if ($resourcesToRemove) {
+        $maximumRetries = '${{ parameters.maximumRemovalRetries }}' -as [int]
+        $currentRety = 0
```
Comment (MariusStorhaug) [inline utilities/resourceRemoval/Remove-vWan.ps1:69]: currentRet*r*y, missing a 'r'.

File: utilities/resourceRemoval/Remove-DeployedModule.ps1 [hunk 4]
```diff
@@ -48,41 +42,26 @@ function Remove-DeployedModule {
...
         ## Process Removal ##
         #####################
-        if ($deploymentScope -eq 'subscription') {
+        if ([String]::IsNullOrEmpty($resourceGroupName)) {
             Write-Verbose 'Handle subscription level removal'
-            $resourceGroupToRemove = Get-AzResourceGroup -Tag @{ removeModule = $moduleName }
+
+            $retryCount = 1
+            $retryLimit = 40
+            $retryWaitInSeconds = 15
```
Comment (MariusStorhaug) [inline utilities/resourceRemoval/Remove-DeployedModule.ps1:53]: Suggestion: Put these in parameters so they can be overridden.

File: utilities/resourceRemoval/Remove-DeployedModule.ps1 [hunk 5]
```diff
@@ -100,7 +79,16 @@ function Remove-DeployedModule {
             }
         } else {
             Write-Verbose 'Handle resource group level removal'
-            $resourcesToRemove = Get-AzResource -Tag @{ removeModule = $moduleName } -ResourceGroupName $resourceGroupName
+
+            $retryCount = 1
+            $retryLimit = 40
+            $retryWaitInSeconds = 15
```
Comment (MariusStorhaug) [inline utilities/resourceRemoval/Remove-DeployedModule.ps1:85]: Repeated code? Again, take these from parameters in my opinion.

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]
Some comments. Also:
- Missing docs on some PowerShell functions.
- Missing subscription and managementGroup scope for removal (if im not mistaken). (Guess this is for the next version.

File: utilities/resourceRemoval/Remove-vWan.ps1 (hunk 1 above)
Comment (MrMCake) [reply utilities/resourceRemoval/Remove-vWan.ps1]: done

File: utilities/resourceRemoval/Remove-NetAppAccount.ps1 (hunk 2 above)
Comment (MrMCake) [reply utilities/resourceRemoval/Remove-NetAppAccount.ps1]: Agreed. I'm currently working on that & aligning the approaches (for resources) too

File: utilities/resourceRemoval/Remove-vWan.ps1 (hunk 3 above)
Comment (MrMCake) [reply utilities/resourceRemoval/Remove-vWan.ps1]: fixed

File: utilities/resourceRemoval/Remove-DeployedModule.ps1 (hunk 4 above)
Comment (MrMCake) [reply utilities/resourceRemoval/Remove-DeployedModule.ps1]: done

File: utilities/resourceRemoval/Remove-DeployedModule.ps1 (hunk 5 above)
Comment (MrMCake) [reply utilities/resourceRemoval/Remove-DeployedModule.ps1]: done

**Review by MariusStorhaug — APPROVED** [review]
All comments seem to be solved. Green pipelines which are also a good indicator of working solution. Lets see if this solves our faulty workflows and removal steps! Approved.

**Pull request merged**